### PR TITLE
Add Haskell support via Stack / Cabal

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

/cc @stevana

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/644)

### CODE REVIEW

File: analyzer/src/main/kotlin/managers/Stack.kt [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        )
+
+        // Stack does not support lock files, so hard-code "allowDynamicVersions" to "true".
+        return ProjectAnalyzerResult(true, project,
+                allPackages.map { it.toCuratedPackage() }.toSortedSet())
+    }
+
+    private fun buildDependencyTree(parentName: String,
+                                    childMap: Map<String, List<String>>, versionMap: Map<String, String>,
+                                    allPackages: SortedSet<Package>, dependencies: SortedSet<PackageReference>) {
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Stack.kt]: Should we start using this proposed formatting?
https://kotlinlang.org/docs/reference/coding-conventions.html#function-formatting

File: analyzer/src/main/kotlin/managers/Stack.kt [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        childMap[parentName]?.forEach { childName ->
+            val pkg = Package(
+                    id = Identifier(
+                            provider = "Stack",
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Stack.kt]: In other classes we use `toString()`.

File: analyzer/src/main/kotlin/managers/Stack.kt [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+        val scopes = sortedSetOf(
+                Scope("external", true, externalDependencies),
+                Scope("test", false, testDependencies),
+                Scope("bench", false, benchDependencies)
+        )
+
+        val project = Project(
+                id = Identifier(
+                        provider = "Hackage",
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Stack.kt]: Why not "stack"?

File: analyzer/src/main/kotlin/managers/Stack.kt [hunk 4]
```diff
@@ -154,4 +194,135 @@ class Stack : PackageManager() {
...
+                .build()
+
+        return OkHttpClientHelper.execute(Main.HTTP_CACHE_PATH, pkgRequest).use { response ->
+            val body = response.body()?.string()?.trim()
+
+            if (response.code() != HttpURLConnection.HTTP_OK || body.isNullOrEmpty()) {
+                log.warn { "Unable to retrieve Hackage meta-data for package '${pkg.id}'." }
+                if (body != null) {
+                    log.warn { "Response was '$body'." }
+                }
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Stack.kt]: Would be nice to print the response code / status line in this case.

File: analyzer/src/main/kotlin/managers/Stack.kt (hunk 1 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Stack.kt]: I basically have nothing against that style, it's just not how the IDE formats the code by default, which makes it a bit inconvenient to use.

When using that style, do you put *each* argument on a single line, even if multiple shorter arguments would fit into a single line?

File: analyzer/src/main/kotlin/managers/Stack.kt (hunk 3 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Stack.kt]: Done.

File: analyzer/src/main/kotlin/managers/Stack.kt [hunk 5]
```diff
@@ -84,47 +93,78 @@ class Stack : PackageManager() {
...
-                        version = ""
-                ),
-                definitionFilePath = VersionControlSystem.getPathToRoot(definitionFile) ?: "",
-                declaredLicenses = sortedSetOf(),
-                vcs = VcsInfo.EMPTY,
-                vcsProcessed = processProjectVcs(workingDir, homepageUrl = ""),
-                homepageUrl = "",
-                scopes = scopes
-        )
+        // Enrich the package templates with additional meta-data from PyPI.
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Stack.kt]: Done.

File: analyzer/src/main/kotlin/managers/Stack.kt (hunk 1 above)
Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/Stack.kt]: Yes, one argument per line. But we haven't used this style so far, so we could also refactor the code later if we want to use it.

**Pull request merged**